notebooks/LeNet-example.ipynb

```diff
@@ -0,0 +1,555 @@
+   "metadata": {},
+   "source": [
+    "<a id=\"2\"></a>\n",
+    "## 2. Creating TorchScript modules\n",
+    "\n",
+    "Here we create two submodules for a feature extractor and a classifier and stitch them together in a single LeNet module. In this case this is overkill but modules give us granular control over our program including where we decide to optimize and where we don't. It is also the unit that the TorchScript compiler operates on. So you can decide to only convert/optimize the feature extractor and leave the classifier in standard PyTorch or you can convert the whole thing. When compiling your module to TorchScript, there are two paths: Tracing and Scripting.  "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import torch \n",
+    "from torch import nn\n",
+    "import torch.nn.functional as F\n",
+    "\n",
+    "class LeNetFeatExtractor(nn.Module):\n",
+    "    def __init__(self):\n",
+    "        super(LeNetFeatExtractor, self).__init__()\n",
+    "        self.conv1 = nn.Conv2d(1, 6, 3)\n",
+    "        self.conv2 = nn.Conv2d(6, 16, 3)\n",
+    "\n",
+    "    def forward(self, x):\n",
+    "        x = F.max_pool2d(F.relu(self.conv1(x)), (2, 2))\n",
+    "        x = F.max_pool2d(F.relu(self.conv2(x)), 2)\n",
+    "        return x\n",
+    "\n",
+    "class LeNetClassifier(nn.Module):\n",
+    "    def __init__(self):\n",
+    "        super(LeNetClassifier, self).__init__()\n",
+    "        self.fc1 = nn.Linear(16 * 6 * 6, 120)\n",
+    "        self.fc2 = nn.Linear(120, 84)\n",
+    "        self.fc3 = nn.Linear(84, 10)\n",
+    "\n",
+    "    def forward(self, x):\n",
+    "        x = torch.flatten(x,1)\n",
+    "        x = F.relu(self.fc1(x))\n",
+    "        x = F.relu(self.fc2(x))\n",
+    "        x = self.fc3(x)\n",
+    "        return x\n",
+    "\n",
+    "class LeNet(nn.Module):\n",
+    "    def __init__(self):\n",
+    "        super(LeNet, self).__init__()\n",
+    "        self.feat = LeNetFeatExtractor()\n",
+    "        self.classifer = LeNetClassifier()\n",
+    "\n",
+    "    def forward(self, x):\n",
+    "        x = self.feat(x)\n",
+    "        x = self.classifer(x)\n",
+    "        return x\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let us define a helper function to benchmark a model."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import time\n",
+    "import numpy as np\n",
+    "\n",
+    "def benchmark(model, input_shape=(1024, 1, 32, 32), dtype='fp32', nwarmup=50, nruns=10000):\n",
+    "    input_data = torch.randn(input_shape)\n",
+    "    input_data = input_data.to(\"cuda\")\n",
+    "    if dtype=='fp16':\n",
+    "        input_data = input_data.half()\n",
+    "        \n",
+    "    for _ in range(nwarmup):\n",
+    "        results = model(input_data)\n",
```

Comment:
Do we need to use pytorch benchmarking methodology here? i.e. setting cudnn benchmark etc?

Reply:
I'm using a super simple benchmarking procedure we adopted for TensorFlow-TensorRT. Quite basic. I think ultimately the user will do their own end to end benchmark with real data etc..,  I think we can keep it simple here.

Reply:
Ok, I think we should at least disclose that it isn't a formal benchmark of pytorch, I know that not having settings like `cudnn_benchmark = true` can lead to some misleading results including potentially seeming faster that TRT. I can also share some benchmark code we have if you want to extend what is here, it shouldnt be to much more complicated.

Reply:
sounds good, pls point me to that. thanks.

Reply:
modified benchmark procedure now. Pls check